drivers/video/video_emul_rx.c

```diff
@@ -0,0 +1,327 @@
+	struct video_format *fmt = &data->fmt;
+	struct video_buffer *vbuf = vbuf;
+
+	LOG_DBG("Queueing a frame of %u bytes in format %x %ux%u", fmt->pitch * fmt->height,
+		fmt->pixelformat, fmt->width, fmt->height);
+
+	while ((vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT)) != NULL) {
+		vbuf->bytesused = fmt->pitch * fmt->height;
+		vbuf->line_offset = 0;
+
+		LOG_DBG("Inserting %u bytes into buffer %p", vbuf->bytesused, vbuf->buffer);
+
+		/* Simulate the MIPI/DVP hardware transferring image data from the imager to the
+		 * video buffer memory using DMA. The vbuf->size is checked in emul_rx_enqueue().
+		 */
+		memcpy(vbuf->buffer, conf->source_dev->data, vbuf->bytesused);
+
+		/* Once the buffer is completed, submit it to the video buffer */
+		k_fifo_put(&data->fifo_out, vbuf);
+	}
+}
+
+static int emul_rx_enqueue(const struct device *dev, enum video_endpoint_id ep,
+			   struct video_buffer *vbuf)
+{
+	struct emul_rx_data *data = dev->data;
+	struct video_format *fmt = &data->fmt;
+
+	/* Can only enqueue a buffer to get data out, data input is from hardware */
+	if (ep != VIDEO_EP_OUT && ep != VIDEO_EP_ALL) {
+		return -EINVAL;
+	}
+
+	if (vbuf->size < fmt->pitch * fmt->height) {
+		LOG_ERR("Buffer too small for a full frame");
+		return -ENOMEM;
+	}
+
+	/* The buffer has not been filled yet: flag as emtpy */
+	vbuf->bytesused = 0;
+
+	/* Submit the buffer for processing in the worker, where everything happens */
+	k_fifo_put(&data->fifo_in, vbuf);
+	k_work_submit(&data->work);
+
+	return 0;
+}
+
+static int emul_rx_dequeue(const struct device *dev, enum video_endpoint_id ep,
+			   struct video_buffer **vbufp, k_timeout_t timeout)
+{
+	struct emul_rx_data *data = dev->data;
+
+	/* Can only dequeue a buffer to get data out, data input is from hardware */
+	if (ep != VIDEO_EP_OUT && ep != VIDEO_EP_ALL) {
+		return -EINVAL;
+	}
+
+	/* All the processing is expected to happen in the worker */
+	*vbufp = k_fifo_get(&data->fifo_out, timeout);
+	if (*vbufp == NULL) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+static int emul_rx_flush(const struct device *dev, enum video_endpoint_id ep, bool cancel)
+{
+	struct emul_rx_data *data = dev->data;
+	struct k_work_sync sync;
+
+	/* Can only flush the buffer going out, data input is from hardware */
+	if (ep != VIDEO_EP_OUT && ep != VIDEO_EP_ALL) {
+		return -EINVAL;
+	}
+
+	if (cancel) {
+		struct video_buffer *vbuf;
+
+		/* First, stop the hardware processing */
+		emul_rx_stream_stop(dev);
+
+		/* Cancel the jobs that were not running */
+		k_work_cancel(&data->work);
+
+		/* Flush the jobs that were still running */
+		k_work_flush(&data->work, &sync);
+
+		/* Empty all the cancelled items */
+		while ((vbuf = k_fifo_get(&data->fifo_in, K_NO_WAIT))) {
+			k_fifo_put(&data->fifo_out, vbuf);
+		}
+	} else {
+		/* Process all the remaining items from the queue */
+		k_work_flush(&data->work, &sync);
+	}
+
+	return 0;
+}
+
+static const struct video_driver_api emul_rx_driver_api = {
```

Comment:
```suggestion
static DEVICE_API(video, emul_rx_driver_api) = {
```

Reply:
Done as well, thank you for the reminder.